futures-util/src/stream/buffer_unordered.rs

```diff
@@ -76,45 +98,41 @@ impl<S> BufferUnordered<S>
 }
 
 impl<S> Stream for BufferUnordered<S>
-    where S: Stream,
-          S::Item: IntoFuture<Error=<S as Stream>::Error>,
+where
+    S: Stream,
+    S::Item: Future,
 {
-    type Item = <S::Item as IntoFuture>::Item;
-    type Error = <S as Stream>::Error;
+    type Item = <S::Item as Future>::Output;
 
-    fn poll_next(&mut self, cx: &mut task::Context) -> Poll<Option<Self::Item>, Self::Error> {
+    fn poll_next(mut self: PinMut<Self>, cx: &mut task::Context) -> Poll<Option<Self::Item>> {
         // First up, try to spawn off as many futures as possible by filling up
         // our slab of futures.
         while self.queue.len() < self.max {
-            let future = match self.stream.poll_next(cx)? {
-                Async::Ready(Some(s)) => s.into_future(),
-                Async::Ready(None) |
-                Async::Pending => break,
-            };
-
-            self.queue.push(future);
+            match self.stream().poll_next(cx) {
+                Poll::Ready(Some(future)) => self.queue().push(future),
+                Poll::Ready(None) | Poll::Pending => break,
+            }
         }
 
-        // Try polling a new future
-        if let Some(val) = ready!(self.queue.poll_next(cx)) {
-            return Ok(Async::Ready(Some(val)));
+        // Attempt to pull the next value from the queue
+        if let Some(val) = ready!(PinMut::new(self.queue()).poll_next(cx)) {
+            return Poll::Ready(Some(val));
         }
```

Comment:
Without the `ready!` macro it could look like this:
```rust
match PinMut::new(self.queue()).poll_next(cx) {
    x @ Poll::Pending => return x,
    x @ Poll::Ready(Some(_)) => return x,
    Poll::Ready(None) => {},
}
``` 
This is IMO way easier to understand what's going

Reply:
Done.